language/move-prover/abigen/src/abigen.rs

```diff
@@ -0,0 +1,189 @@
+}
+
+impl Default for AbigenOptions {
+    fn default() -> Self {
+        Self {
+            compiled_script_directory: ".".to_string(),
+            output_directory: "abi".to_string(),
+        }
+    }
+}
+
+/// The ABI generator.
+pub struct Abigen<'env> {
+    /// Options.
+    options: &'env AbigenOptions,
+    /// Input definitions.
+    env: &'env GlobalEnv,
+    /// Map from file name to generated script ABI (if any).
+    output: BTreeMap<String, ScriptABI>,
+}
+
+impl<'env> Abigen<'env> {
+    /// Creates a new documentation generator.
+    pub fn new(env: &'env GlobalEnv, options: &'env AbigenOptions) -> Self {
+        Self {
+            options,
+            env,
+            output: Default::default(),
+        }
+    }
+
+    /// Returns the result of ABI generation, a vector of pairs of filenames
+    /// and JSON content.
+    pub fn into_result(mut self) -> Vec<(String, Vec<u8>)> {
+        std::mem::take(&mut self.output)
+            .into_iter()
+            .map(|(path, abi)| {
+                let content = lcs::to_bytes(&abi).expect("ABI serialization should not fail");
+                (path, content)
+            })
+            .collect()
+    }
+
+    /// Generates ABIs for all script modules in the environment (exclusing the dependency set).
+    pub fn gen(&mut self) {
+        for module in self.env.get_modules() {
+            if module.is_script_module() && !module.is_dependency() {
+                let mut path = PathBuf::from(&self.options.output_directory);
+                path.push(
+                    PathBuf::from(module.get_source_path())
+                        .with_extension("abi")
+                        .file_name()
+                        .expect("file name"),
+                );
+
+                match self.compute_abi(&module) {
+                    Ok(abi) => {
+                        self.output.insert(path.to_string_lossy().to_string(), abi);
+                    }
+                    Err(error) => panic!(
+                        "Error while processing script file {:?}: {}",
+                        module.get_source_path(),
+                        error
+                    ),
+                }
+            }
+        }
+    }
+
+    /// Compute the ABI of a script module.
+    fn compute_abi(&self, module_env: &ModuleEnv<'env>) -> anyhow::Result<ScriptABI> {
+        let symbol_pool = module_env.symbol_pool();
+        let func = match module_env.get_functions().next() {
+            Some(f) => f,
+            None => bail!("A script module should define a function."),
+        };
+        let name = symbol_pool.string(func.get_name()).to_string();
+        let doc = func.get_doc().to_string();
+        let code = self.load_compiled_bytes(&module_env)?.to_vec();
+        let ty_args = func
+            .get_named_type_parameters()
+            .iter()
+            .map(|ty_param| {
+                TypeArgumentABI::new(symbol_pool.string(ty_param.0).to_string().to_lowercase())
```

Comment:
I don't think this is a major issue, but if we have two types `Cointype` and `CoinType` they'd both be mapped to `cointype` in this transformation, while also being viewed as different type parameters in Move. I doubt we'd encounter it normally, but wondering if going to snake case (converting upper to `_lower`) might be better? (e.g. the two examples would be converted to `coin_type` and `cointype`)

Reply:
I agree snake case would be better. Let me see if I can find a library for this.

Reply:
For name collisions, I need to think about it more. I don't know if we can realistically achieve a truly non-failing (aka "monkey proof") code gen.

Reply:
Agreed on the "monkey proof" code gen. I trust your judgement on this though, so I'm interested to see what you come up with! :)